cirq/circuits/frozen_circuit.py

```diff
@@ -85,13 +89,19 @@ def moments(self) -> Sequence['cirq.Moment']:
     def device(self) -> devices.Device:
         return self._device
 
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
     def __hash__(self):
-        return hash((self.moments, self.device))
+        return hash((self.moments, self.device, self.name))
```

Comment:
IIUC, this poses a risk that a user could add duplicate definitions of circuits to the serialization by providing them with unique names. Maybe it's future work, but should there be a check somewhere to prevent this?

Reply:
We could check during serialization if two `FrozenCircuit`s are equal apart from their names, I guess - unless someone really likes reading JSON-formatted circuits, there probably isn't a use case for identical circuits with different names.

(Aside: thanks to #3673, hash values are not required for concise serialization)